### PR TITLE
[NAE-1478] ActionDelegate functions overloading

### PR DESCRIPTION
**Pull request by renczesstefan**
# Description

Implemented overloaded function for PDF generator in ActionDelegate.

Implements [NAE-1478]

## Dependencies

No new dependencies were introduced.

### Third party dependencies

No new dependencies were introduced.

### Blocking Pull requests

There are no dependencies on other PR.

## How Has Been This Tested?

This was tested manually generating PDF.

### Test Configuration

Netgrif Backend PR Config

| Name                | Tested on |
|---------------------| --------- |
| OS                  |   macOS Monterey 12.2.1        |
| Runtime             |  Java 11        |
| Dependency Manager  |  Maven 3.8.4         |
| Framework version   |  Spring Boot 2.6.2        |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes have been checked, personally or remotely, with @petrovicLubos 
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have resolved all conflicts with the target branch of the PR
- [x] I have updated and synced my code with the target branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes:
    - [x] Lint test
    - [x] Unit tests
    - [x] Integration tests
- [x] I have checked my contribution with code analysis tools:
    - [x] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [x] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [x] I have made corresponding changes to the documentation:
    - [x] Developer documentation
    - [x] User Guides
    - [x] Migration Guides

[NAE-1478]: https://netgrif.atlassian.net/browse/NAE-1478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by machacjozef — DISMISSED** [review]

**Review by minop — APPROVED** [review]

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 1]
```diff
@@ -776,32 +786,36 @@ class ActionDelegate {
...
-    void generatePDF(String transitionId, String fileFieldId) {
+    void generatePDF(Transition t, FileField field) {
+        this.generatePDF(t.getStringId(), field.importId)
+    }
+
+    void generatePDF(String transitionId, String fileFieldId, Case fromCase = useCase, Case saveToCase = useCase) {
         PdfResource pdfResource = ApplicationContextProvider.getBean(PdfResource.class) as PdfResource
         String filename = pdfResource.getOutputDefaultName()
-        String storagePath = pdfResource.getOutputFolder() + File.separator + useCase.stringId + "-" + fileFieldId + "-" + pdfResource.getOutputDefaultName()
+        String storagePath = pdfResource.getOutputFolder() + File.separator + saveToCase.stringId + "-" + fileFieldId + "-" + pdfResource.getOutputDefaultName()
```
Comment (minop) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:796]: the PDF methods seem very similar, which is amplified by the fact that the lines are displayed twice in the diff. In my opinion, at least the `storagePath` creation should be abstracted into a single method call in all implementations.

**Review by minop — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 2]
```diff
@@ -749,6 +753,12 @@ class ActionDelegate {
         return mapData(addGetDataOutcomeToOutcomesAndReturnData(dataService.getData(task, useCase)))
     }
 
+    Map<String, Field> getData(String taskId) {
+        Task task = taskService.findById(taskId)
+        def useCase = workflowService.findOne(task.caseId)
+        return mapData(addGetDataOutcomeToOutcomesAndReturnData(dataService.getData(task, useCase)))
```
[the remark this answers is not in the document]
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:759]: There is no function getData(Task, Case).

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 1 above)
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy]: Corrected

**Review by minop — APPROVED** [review]

**Review by minop — DISMISSED** [review]

**Comment by renczesstefan** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=netgrif_application-engine&pullRequest=40)
> 
> [Bug](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=netgrif_application-engine&pullRequest=40&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=netgrif_application-engine&pullRequest=40&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=netgrif_application-engine&pullRequest=40&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=netgrif_application-engine&pullRequest=40&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=netgrif_application-engine&pullRequest=40&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=netgrif_application-engine&pullRequest=40&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=netgrif_application-engine&pullRequest=40&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=netgrif_application-engine&pullRequest=40&metric=new_duplicated_lines_density&view=list)

The code coverage fail is because of ActionDelegate functions. Tests cannot be implemented directly to ActionDelegate functions, but the same functions are implemented in test files.

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 3]
```diff
@@ -699,7 +699,11 @@ class ActionDelegate {
     }
 
     SetDataEventOutcome setData(Task task, Map dataSet) {
-        return addSetDataOutcomeToOutcomes(dataService.setData(task.stringId, ImportHelper.populateDataset(dataSet)))
+        setData(task.stringId, dataSet)
```
Comment (mladoniczky) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy]: Please explicitly define the `return` statement.

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy [hunk 4]
```diff
@@ -779,32 +789,50 @@ class ActionDelegate {
         return userService.loggedUser
     }
 
-    void generatePDF(String transitionId, String fileFieldId) {
+    void generatePDF(Transition t, FileField field) {
+        this.generatePDF(t.getStringId(), field.importId)
+    }
+
+    String buildStoragePath(String outputFolder, String outputName, Case aCase, String fileFieldId) {
```
Comment (mladoniczky) [inline src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy:796]: This is duplicate implementation of the method `FileFieldValue.getPath` .

**Review by mladoniczky — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 4 above)
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy]: This function is now removed and the suggested method is used.

File: src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy (hunk 3 above)
Comment (renczesstefan) [reply src/main/groovy/com/netgrif/application/engine/petrinet/domain/dataset/logic/action/ActionDelegate.groovy]: Done.

**Review by mladoniczky — APPROVED** [review]

**Pull request merged**